<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>Help</title>
  </head>
  <body>
    <h1>Contents</h1>
    <ul>
      <li><a href="#Instructions_For_Use">Instructions For Use</a></li>
      <ul>
        <li><a href="#Password">Password</a></li>
        <li><a href="#Loading_Transcriptomics_Data">Loading
            Transcriptomics Data</a></li>
        <li><a href="#Explore_Transcriptomics_By_Probe_tab">Explore By
            Gene<br>
          </a></li>
        <ul>
          <li><a href="#Select_tab_-_probes_">Select Genes</a></li>
          <ul>
            <li><a href="#Rows_Limit_Cap">Rows Limit Cap</a><br>
            </li>
          </ul>
          <ul>
            <li><a href="#1._Regex_-_keyword_search_">1. Regex - keyword
                search</a></li>
            <li><a href="#2_and_3:_Values_In_Range_">2 &amp; 3: Values
                In Range and Rows In Range</a></li>
            <ul>
              <li><a href="#Sort_descending">Sort descending</a><br>
              </li>
              <li><a href="#Gene_Averages">Gene Averages</a></li>
              <li><a href="#2._Values_In_Range">2. Values In Range</a></li>
              <ul>
                <li><a href="#Single_TreatmentTime_combination">Single
                    Treatment~Time combination</a></li>
                <li><a href="#At_any_time_for_the_treatment_selected">At
                    any time for the treatment selected</a><br>
                </li>
                <li><a href="#Matching_kinetics_for_one_treatment">Matching&nbsp;










                    kinetics for one treatment over different times</a></li>
              </ul>
              <li><a href="#3._Rows_In_Range_">3. Rows In Range</a><br>
              </li>
            </ul>
          </ul>
          <li><a href="#Selected_Probes_tab">Selected Genes<br>
            </a></li>
          <li><a href="#Probes:Series_tab_">Gene:Series</a></li>
          <li><a href="#Genes-Modules_tab_">Genes-&gt;Modules</a></li>
          <li><a href="#Modules_tab_">Modules</a></li>
          <li><a href="#Modules-Genes_Probes_tab">Modules-&gt;Genes</a></li>
          <li><a href="#Modules:Series_Probes_tab">Modules:Series</a></li>
        </ul>
        <li><a href="#Explore_Transcriptomics_By_Module_tab">Explore
            Transcriptomics By Module</a></li>
        <ul>
          <li><a href="#Select_modules_tab_">Select modules</a></li>
          <li><a href="#Selected_Modules_Modules_tab_">Selected Modules</a></li>
          <li><a href="#Modules-Genes_Modules_tab">Modules-&gt;Genes</a></li>
          <li><a href="#Modules:Series_Modules_tab">Modules:Series</a></li>
        </ul>
        <li><a href="#Transcriptomics_Lookup_tab_">Transcriptomics
            Lookup</a></li>
        <ul>
          <li><a href="#Gene_Lookup_tab">Gene Lookup</a></li>
          <li><a href="#Module_Lookup_tab">Module Lookup</a></li>
        </ul>
        <li><a href="#Explore_Cytokines">Explore Cytokines</a></li>
        <ul>
          <li><a href="#Plot_Cytokines">Plot Cytokines</a></li>
        </ul>
        <li><a href="#Explore_Cells">Explore Cells</a><br>
        </li>
        <ul>
        </ul>
      </ul>
    </ul>
    <h1><a name="Instructions_For_Use"></a>Instructions For Use</h1>
    <h2><a name="Password"></a> Password</h2>
    When running on the shinyapps.io server the software requires a
    password. This is to protect the datasets. Please contact
    d.j.lewis@surrey.ac.uk to be given the password. Enter the password
    and click the button to unlock the menus. When running locally no
    password is required. <br>
    <h2><a name="Loading_Transcriptomics_Data"></a> Loading
      Transcriptomics Data</h2>
    You must load a transcriptomics dataset before you can do anything
    with transcriptomics data. The transcriptomics datasets are already
    uploaded to the server. Select a dataset. Click <b>Load Dataset</b>.
    All existing plots and data tables will be cleared. Once you click
    "Load" the name of the dataset will appear below the tabs, and two
    more tabs will appear:<br>
    <ul>
      <li><a href="#Explore_Transcriptomics_By_Probe_tab">Explore By
          Gene<br>
        </a></li>
      <li><a href="#Explore_Transcriptomics_By_Module_tab">Explore By
          Module</a></li>
    </ul>
    <h2><a name="Explore_Transcriptomics_By_Probe_tab"></a>Explore By
      Gene<br>
    </h2>
    Select this tab to explore the datasets using individual spots in
    the array.<br>
    <br>
    <i>Often the rows are are referred to as "probes" but I prefer to
      call them "spots" as each row in the data has an unique ID and
      refers to a particular spot on the chip by its column-row
      position, so you can compare exact spots on the array across
      treatments and times. In addition, a probe (nucleic acid sequence)
      may be represented on multiple replicate spots. Therefore I use
      ProbeName to refer to probes and Spots to refer to rows in the
      expression array matrices.</i> <br>
    <h3><a name="Select_tab_-_probes_"></a>Select Genes<br>
    </h3>
    <font size="+2" color="#009900"><b> Filtering Spots Rows</b></font><br>
    You must select at least one of the numbered filters, and you must
    click the
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <b>Apply Filters</b> button before any filters or selections are
    applied, and again after any changes. This is to speed-up
    calculations. Once you click
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <b>Apply Filters</b>, several other tabs will appear to allow
    visualisation of the filtered genes and modules.<br>
    <br>
    <u><b>Filters are applied one after another, left to right: each
        filter applies its filter and then hands the reduced set of rows
        to the next filter.</b></u> <br>
    First the <b>keyword</b> search will filter rows based on
    TRUE/FALSE test of finding a keyword in the annotation columns.
    Filtered rows are then passed to the <b>Values In Range</b> filter
    for further reduction. The <b>Values In Range&nbsp;</b> filter is
    applied only to the columns containing data for one treatment
    (vaccine or placebo) selected in a menu. However, the filter may be
    applied either to a single column (one time point) or to all
    columns, with a TRUE/FALSE result as to whether the row has a value
    within the range. An option to vary the range by column is also
    available (kinetics filtering). Filtered rows are then sorted
    numerically, either using the values in one column (using a time
    point selected in a menu) or across all columns for that treatment.
    The sorted rows are passed to the <b>Rows In Range</b> filter which
    can take the N top, bottom or slice of ordered rows. The rows
    remaining after all the filters are used in subsequent charts and
    tables.<br>
    <h5><a name="Rows_Limit_Cap"></a>Rows Limit Cap</h5>
    At the top right of the screen is a box into which you can type an
    upper limit of rows that can be returned by filters. This defaults
    to 100. If a set of filters returns more than this number of rows
    the data will be capped to the value in this box. This acts as a
    safeguard against the app freezing for a long time when an
    indiscriminate filter returns many rows. Note however that the rows
    are capped <b>after sorting.</b><br>
    <h4><a name="1._Regex_-_keyword_search_"></a>1. Regex - keyword
      search<br>
    </h4>
    This is a keyword, free-text search that uses regex, with some
    limitations. Multiple terms can be entered into the box <b>separated






















      by commas</b> and are OR'd (search returns as many of those terms
    as it can find, treating each term separately). Commas are
    interpreted as search term dividers and cannot be 'escaped'. Select
    whether to search gene name or description, spot ID ("Spot"),
    ProbeName or gene SystematicName, which are all standardised between
    datasets.<br>
    <br>
    Spaces will be stripped-out automatically in all searches except
    Gene Description, so you can easily paste in lists copied from
    tables. Regex is "literal" so if you type <tt>"B cell"</tt> it is
    going to look for <tt>"B cell"</tt> - including the <tt>""</tt>. <br>
    <br>
    You can enter partial terms e.g. interferon finds interferon and
    interferons, as regex looks for those characters anywhere. If you
    select "Whole Word" then interferon will only match interferon. <b>
      Whole Word is important when searching for genes and spots</b> as
    "CCL2" will find CCL2, CCL27 etc unless Whole Word is selected, and
    this is worse for spots as they are numeric. Upper/lower case is
    ignored. Beware including <a href="https://regexr.com"
      target="_blank">special regex characters</a> by mistake.<br>
    <br>
    Keyword search&nbsp; ignores the other search filters, it just finds
    rows in the data set that match the search in the selected
    annotation column, and then hands that reduced set on to the next
    filters.<br>
    <br>
    You can also <b>paste saved search lists</b> into the regex box
    that you downloaded from the <b>Selected Spots tab</b>, or copied
    from one of the <b>Lookup</b> tables. <b>This allows fairly easy
      repetition of searches across datasets</b>.<br>
    <h4><a name="2_and_3:_Values_In_Range_"></a>2 and 3: Values In Range
      and Rows In Range<br>
    </h4>
    <p>In both cases you filter rows based on expression / fold change
      values. You must must pick a&nbsp; <b>treatment</b> from the
      menu, which will map to a set of columns of data in the dataset.
      You must also pick a <b>time</b>, which may be used in the filter
      (see below) and is always used to show box plots for the modules,
      as modules are summarised only for one time point. Time series
      plots will indicate this <b>treatment~time combination</b> with <font
        color="#ff0000">▼▲</font> on the chart.<br>
      <br>
      Max/min expression values will be pre-entered for you into the
      range boxes when selecting a column, using floor() and ceil() and
      so will deliberately exceed the actual values to avoid issues with
      floating-point maths. Click <b>Column</b> to reset to max/min for
      the selected treatment~time column, or <b>Data</b> for max/min of
      whole dataset.</p>
    <h5><a name="Gene_Averages"></a>Gene Averages</h5>
    <p>Spots are arranged in rows - one unique spot per row. However,
      many spots may map to a single probe and many probes to a single
      gene. You can filter by the values of the individual spots or have
      tmodExplorer group the spots by Gene and then calculate <font
        color="#ff0000"><b>Gene Averages</b></font> by which to filter.
      If your main purpose is to identify spots of interest then avoid
      gene averaging which has multiple consequences. However, as the
      modules work in genes, it is probably better to select Gene
      Averages if your main interest is to study modules - as this way
      you increase the number of genes identified if you are restricting
      the number of rows returned. A disadvantage of gene averages is
      that you lose the gene descriptions, which are spot-specific. You
      can copy the <b>Gene list</b> on the <a
href="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/help2.html#Selected_Probes_tab"><b>Selected



          Genes</b></a> tab and use&nbsp; the <a
href="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/help2.html#Gene_Lookup_tab"><b>Gene











          Lookup tab</b></a> to lookup genes by their names and see the
      descriptions.</p>
    <h5><a name="2._Values_In_Range"></a>2. Values In Range</h5>
    <p>This filter operates in one of three ways according to how you
      select <b>Treat~Time</b>, <b>Treat~Kinetics</b>, and <b>All
        Times</b>. Basically, while the treatment selected will always
      be used to filter, the number of time columns that you include can
      be varied:<br>
      <b></b></p>
    <ol>
      <li>Filter by values in the <b>single Treatment~Time combination</b>
        (one dataset column) selected in the menus when <b>Treat~Time</b>
        selected and <b>All Times </b>not selected.<br>
      </li>
      <li>Filter on the values <b>at any time for the treatment
          selected in the menu</b> (all columns for that treatment,
        including days 0 and -1): when <b>Treat~Time</b> and <b>All
          Times</b> selected.<br>
      </li>
      <li>Filter by <b>matching the kinetics for one treatment at
          specified times</b> when <b>Treat~Kinetics</b> is selected
        (different ranges on different days).<br>
      </li>
    </ol>
    <h6><a name="Single_TreatmentTime_combination"></a><b>Single
        Treatment~Time combination</b></h6>
    <p>This is the simplest way to filter by value and uses the
      Treatment~Time combination you selected in the menus to pick one
      column of data to apply the filter to. Spots or Gene averages will
      be selected if the value falls between the Lowest - Highest values
      in the Treatment~Time column you have selected <i>e.g. select
        spots whose value falls in this range on Day 1 after Fluad</i>.
      The purpose is to select spots or genes (when Gene Averages
      selected) of interest in the response to a specific treatment, on
      a specific day. The modules values will be calculated for the same
      day in the summary box plots.<br>
    </p>
    <h6><a name="At_any_time_for_the_treatment_selected"></a>At any time
      for the treatment selected</h6>
    <p>This is a variant of Single Treatment~Time combination, except
      that the time point selected in the menu is ignored (the menu goes
      grey to reflect this) and the max or min value for each row in all
      the columns for that treatment is identified, and then the rows
      are filtered against that value range. Even day 0 and day -1
      values will be used to rank the rows. Max() is used when Sort
      Descending is selected (highest values on top) and min() when sort
      descending is not selected (lowest values on top). Note however
      that the time point selected on the menu is used to present the
      summary box plots for the modules mapping filtered genes, so it is
      worth selecting a meaningful day, or you will have to re-filter.<br>
    </p>
    <h6><a name="Matching_kinetics_for_one_treatment"></a><b>Matching&nbsp;










        kinetics for one treatment over different times</b></h6>
    <p>This is a multi-column version of the <b>Treat~Time</b> filter.
      The <b>All Times</b> option is ignored as you can set which days
      to include individually, as is the time menu (the menu goes grey
      to reflect this). The treatment you have selected in the menu is
      used to restrict the filter to the columns with data for that
      treatment. The purpose is to filter on the shape of the kinetics
      curve of the spot / gene means responses after a specific
      treatment. For example: You may want to identify spots whose
      expression following Fluad is high only on day 1&nbsp; and then
      falls to baseline thereafter; or those that remain high on days 1
      &amp; 2 and then go to baseline; etc&nbsp; - - possibly matching
      the kinetics of different cell types.<br>
    </p>
    <p>A bar chart appears, initially showing the maximum and minimum
      values (dotted lines) for the treatment you selected in the menu,
      and bars ("windows") are created for every time point whose
      heights are set to the maximum - minimum values for the whole
      dataset (all treatments - all times). Do not click "Apply Filters"
      with these windows as it just returns the first N rows of the
      dataset depending on what N value you have set in the <a
        href="#Rows_Limit_Cap">rows limit cap</a> box.<br>
    </p>
    <p>Use the menu to pick a time point, or single-click the * inside
      the bar to edit that day. Enter the lowest : highest values and
      click <b>Set</b>. Click&nbsp;<img src="tag.png" alt="" width="23"
        height="20"> to reset to max/min for selected treatment,
      or&nbsp;<img src="dataset.png" alt="" width="22" height="20"> for
      max/min of whole dataset. <b>Remember to click Set otherwise
        changes will be lost. </b>The plot re-draws the new window. <br>
    </p>
    <p>To have a day ignored (no filter) select <b>Ignore</b>&nbsp; -
      the lowest : highest boxes are hidden and the plot draws an empty
      window set to the dataset min-max to illustrate that all values
      will be accepted. When filtering fold-change data it is probably
      worth ignoring day 0 (always 0 value) to save time. <b>Toggle: </b>double-click




      a window bar on the plot to toggle ignore on / off and save the
      change. You cannot ignore all windows - at least one must be open.<br>
      Click&nbsp;<img src="copy.png" alt="" width="22" height="20"> to
      copy the Ignore/Min/Max settings for the day, and click&nbsp;<img
        src="paste.png" alt="" width="20" height="20"> to past the
      copied setting to a day and save.<b><br>
      </b>To reset all the windows to the max min for the treatment
      selected or whole dataset click&nbsp; <img
src="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/tag.png"
        alt="" width="23" height="20"> or&nbsp;<img
src="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/dataset.png"
        alt="" width="22" height="20">. <br>
      To save the windows profile click <b>Export</b>, and to import a
      saved profile click <b>Import</b>….&nbsp; the app will try hard
      to import any profile even if saved from a different dataset, and
      will warn you if they appear not to match the current dataset.<br>
    </p>
    <h5><a name="3._Rows_In_Range_"></a>3. Rows In Range<br>
    </h5>
    This filter operates on rows that have been <u>ordered</u>
    according to value. You can filter to have top N or bottom N rows,
    or have a slice. If there are fewer rows available after the
    previous selections than you request you will get a cryptic warning.
    <br>
    <br>
    Row restrictions apply to the <b>sorted values</b> in either the <b>Treatment~Time











      combination</b> selected in the menus (single column), or across
    all the columns for the selected treatment if <b>All Times</b> is
    selected (note days 0 and -1 also included). When <b>All Times</b>
    is selected the max() value in the columns is used when <b>Sort
      Descending </b>selected, otherwise the min().<br>
    <b>The fact that the rows are ordered taking into account the time
      points and then filtered</b><b> is extremely important</b>, as the
    output from previous filters (keyword, value, kinetics) will be <b>re-filtered



      based on the values in the selected treatment~time column(s)</b>.
    This may have unexpected effects when filtering by kinetics, so <b>pay</b><b>
      attention to the time selected</b> if you have filtered by <b>kinetics</b>.
    For example if you set up a kinetics profile that filters on days 2
    &amp; 3 but then restrict the number of rows with day 7 selected in
    the treatment~time combination it will pull out the top rows based
    on the expressions on day 7, which is probably not what you want.
    Generally avoid filtering by row order when filtering with kinetics.<br>
    If you have <b>Gene Averages</b> selected this affects the the
    number of spots returned: e.g. if you filter to the first 10 rows
    you may include 10 spots, but if <b>Gene Averages</b> is selected
    you get the top 10 genes which may map to more than 10 spots.<br>
    <h5><a name="Sort_descending"></a><b>Sort descending</b></h5>
    Select<b> Sort descending</b> for highest values in top rows, or
    deselect to have lowest values in top rows (in which case top N rows
    is the N most down-regulated).<br>
    <h3><a name="Selected_Probes_tab"></a>Selected Genes</h3>
    A data table showing the spots or genes (if averaged) that meet the
    filter settings is shown and can be downloaded as a csv file or as a
    PNG picture. The <b>Value</b> shown is the one in the "Column" you
    selected. You may use the search boxes above and below the table to
    further restrict the rows displayed in the table, but this has no
    effect on other tabs or plots. Click "⤓ List" to save a
    comma-separated text file that can be copy-pasted into the regex
    search box to quickly repeat the search. Note a description appears
    when spots are displayed individually, and disappears if "Gene
    Averages" is selected on the <b>Select</b> tab. This is because
    descriptions are specific for probes.<br>
    <h3><a name="Probes:Series_tab_"></a> Genes:Series<br>
    </h3>
    This allows the values of the spots or gene averages selected by the
    filters to be plotted for different treatment ~ time combinations,
    to visualise a time course. Use the slider to change plot height,
    and browser window width to change plot width. Enter a number in the
    box to set the maximum number of panels per row when <b>Split</b>
    is used to individually plot treatments. The selected treatment~time
    column is indicated by <font color="#ff0000">▼▲<font
        color="#000000"> on the plot. </font></font><font
      color="#ff0000"><font color="#000000"> Click or drag over the
        points to identify the spot, and you can copy the gene names and
        spot IDs as a comma-separated list for pasting into keyword
        search boxes. </font> </font>
    <h4> Treatment, Days and Gene/Spot menus<br>
    </h4>
    Click 'All' to enter all available, or select one by one from
    pop-down menu and they appear in the box. Click 'None' to clear. The
    order of the columns as they appear in the box is respected for all
    boxes like this - so the plot axis will retain the sequence you
    enter here, unless it is plotting as an XY scatter using numerical
    values. Use the cursor to move between entries to add or delete a
    column at that point. Select an entry and delete.<br>
    <br>
    <b> <b>Plot button</b>: </b>click this to draw the plot, and again
    after changing any options.<b><br>
      Split</b>: the data are split by treatment using facet_wrap.<br>
    <b>Lines / Boxplot</b>: XY graph of individual spots / genes, or a
    boxplot summarising spot/gene values.<br>
    <b> <b>0 |---- (zero) line</b>: </b>Place a dotted line at Y = 0
    (baseline for Fold Change).<b><br>
      Points</b>, <b>Gridlines, </b><b>Legend</b>: show/hide.<br>
    <b>SEM</b>: Appears when Lines selected - show the SEM of spot
    values mapping that gene as a ribbon<br>
    <b><a name="Gene⫷Probes"></a>Gene⫷Spots</b>: Appears when <font
      color="#ff0000"><b>Gene Averages</b></font> selected, and <b>Lines</b>
    selected to plot. Instead of plotting a single line for the average
    of the spot values mapping individual genes (±SEM), separate lines
    are plotted for each of the spots mapping to a gene - the lines are
    colour-coded to the gene, and different symbols are used. This helps
    to show the behaviour of different spots mapping the same gene.<br>
    <b>Kinetics Filters</b> - if you filtered using <b>2: Filter By
      Value</b> and used kinetics, this option appears to draw
    rectangles for the windows used to filter values at each time point.
    It is instructive to use this when <b>Gene Average</b> was
    selected, in combination with<b> </b><b>Gene⫷Spots</b> to identify
    spots that fall outside the windows but were included by the gene
    mean.
    <h3><a name="Genes-Modules_tab_"></a> Genes-&gt;Modules<br>
    </h3>
    This panel lists the filtered spots or genes and shows which - if
    any - modules contain that gene. Some genes you selected may not be
    included in any modules - pay attention to the Rank column as it
    assists with understanding which rows are missing because they do
    not match any modules.
    <h3><a name="Modules_tab_"></a> Modules<br>
    </h3>
    A boxplot summary of the values of the individual spots contained in
    the modules associated with the selected spots or genes. <b>Note:</b>
    the values are from the treatment-time combination selected in the
    menus in <b>Select Genes</b> - to see a different time or treatment
    you must change the filters and re-apply. <br>
    Chart is plotted automatically. Use the slider to change plot
    height, and browser window width to change plot width. If you change
    any of the criteria on the <b>Select</b> tab this plot will
    re-draw. <br>
    <br>
    Show or hide the <b>legend</b> and <b>0 |---- (zero) line</b>, or
    have ggplot2 graphics which are a bit nicer but slower. ggplot2
    calculates the dimensions of the box slightly differently to base
    graphics.<br>
    <br>
    Select "<b>Include Selected As Module</b>" and the genes selected by
    the filters can also be plotted as a "pseudo-module", for comparison
    with&nbsp; the tmod modules.<br>
    "<b>Module</b>" plots one box per module, and "<b>Title</b>" groups
    the modules&nbsp; by their functional / descriptive category titles
    which may result in fewer boxes.<br>
    The table gives the summary statistics and the number of spots
    mapped to the modules' genes (N).<br>
    <h3><a name="Modules-Genes_Probes_tab"></a> Modules-&gt;Genes<br>
    </h3>
    This expands the plot on the <a href="#Modules_tab_"><b>Modules</b></a>
    tab&nbsp; to allow an inspection inside the modules, one by one.
    Select a module from the menu to see the expression of its component
    genes. As each gene may have many spots a summary boxplot of the
    expression values of the spots mapping to the genes inside the
    selected module is plotted automatically. Note that a module may
    also contain genes that are not in the selection created by the <b>Select</b>
    tab. Genes within a module that are in your selection are denoted by
    the ︎► symbol next to the y axis, and in the data table. You may be
    surprised by how few genes in a module overlap with your selected
    genes. <br>
    <br>
    Genes listed for the module but absent from the dataset spots, are
    labelled "Missing" on the plot, which can be toggled on-off. Show or
    hide the <b>legend</b> and <b>0 |---- (zero) line</b> , or use <b>ggplot2</b>
    instead of base graphics.&nbsp; ggplot2 uses a geom_boxplot with
    default values. Unlike the Selected Spots tab which shows the gene
    description in the table, the <b>Module-&gt;Genes</b> tab does not,
    as the descriptions refer to the <i>spots</i>, and here gene
    summaries are being shown. Use the slider to change plot height, and
    browser window width to change plot width.
    <h3><a name="Modules:Series_Probes_tab"></a>Modules:Series</h3>
    While <b>Modules</b> and <b>Modules-&gt;Genes</b> only show the
    values of spots at one selected treatment ~ time combination, <b>Modules:Series</b>
    allows the values of the associated modules to be plotted for all
    treatment ~ time combinations as a time course. To avoid delays when
    selecting options, nothing is plotted or re-plotted until the '<b>Plot</b>'
    button is clicked. Use the slider to change plot height, and browser
    window width to change plot width. Enter a number in the box for
    maximum panels per row when <b>Split</b> selected.
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    Click or drag over the points to identify the module
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <font color="#ff0000"><font color="#000000"> and you can copy the
        module names as a comma-separated list for pasting into keyword
        search boxes</font></font>. <br>
    <br>
    Select treatment ~ time combinations from the menus. Select modules
    from the <b>Modules</b> menu which lists ones mapped to your
    selected spots or genes, as shown in <b>Modules</b>. Use <b>All</b>
    or <b>None</b> buttons to add all, or clear selections.<br>
    <br>
    <b>Boxplot</b> summarises the values of spots within the module.<br>
    <b>Lines</b> plots a line graph of the mean value of spots in the
    modules<b>.<br>
      SEM</b>, <b>Points</b> and <b>Gridlines </b>appear when <b>Lines</b>
    is selected allowing an SEM ribbon,&nbsp; points and gridlines to be
    plotted.<br>
    <b>Split</b> - uses facet_wrap to separate treatments into different
    panels. <br>
    Show / hide the <b>legend</b> and <b>0 |---- (zero</b>) line.<br>
    <h3> </h3>
    <h2><a name="Explore_Transcriptomics_By_Module_tab"></a>Explore
      Transcriptomics By Module tab<br>
    </h2>
    The concept is similar to the <a
      href="#Explore_Transcriptomics_By_Probe_tab"><b>Explore By Spot</b></a>
    tab. However, here a data table is created in which the rows are
    sorted by the <b> mean or median value of the Blood Transcriptional
      Modules</b> as expressed in the selected treatment ~ time
    combination: e.g. the median expression of the spots within a given
    module on day 7 after Boostrix. It is important to note that tmod
    modules contain only a fraction of all the genes in the array, but
    this tab allows for an exploration of responses of the modules
    included in tmod.<br>
    <h3><a name="Select_modules_tab_"></a> Select modules tab<br>
    </h3>
    In this case each row of the data table is the mean or median
    expression value or the fold change from baseline for one Blood
    Transcriptional Module, rather than a spot expression value or fold
    change. Row sorting is based on the mean or median expression (which
    you can select) of the spots within the module. Sort descending has
    the highest expression values in top rows.<br>
    <br>
    See the help for <b>Explore By Spot</b> <a
      href="#Select_tab_-_probes_">Select</a> tab for description of how
    to use <b>regex Keyword</b>, <b>Values Within Range</b> and <b>Row






      Numbers Within Range</b> filters. Time series plots will indicate
    the sorted treatment~time column with <font color="#ff0000">▼▲</font>.
    The <b>Values Within Range</b> filter only applies the filter to
    one Treatment~Time combination (one column of data) as <b>All Days</b>
    and <b>Kinetics</b> filters are not available.<br>
    <br>
    <b>Regex</b> searches either the <b>Title</b> (which is a
    description of the module, e.g. "plasma cells") or <b>Module</b>
    name (its designation e.g. DC.M4.11). Spaces will automatically be
    stripped from Module searches but retained for Title searches. <br>
    You can also <b>paste saved search lists</b> into the regex box
    that you downloaded from the <b>Selected Modules tab</b> - be sure
    to match Title and Module Name. This allows fairly easy repetition
    of searches across datasets.<br>
    You must click the <b>Apply Filters</b> button to perform the
    selection or after any changes to options. Clicking the button
    reveals <b>Selected Modules</b>,&nbsp; <b>Modules-&gt;Genes</b>
    and <b>Modules:Series</b> tabs to visualise the modules.<br>
    <h3><a name="Selected_Modules_Modules_tab_"></a> Selected Modules
      (Modules) tab<br>
    </h3>
    This tab plots a boxplot summary of the spot expressions mapping the
    genes in the modules you selected. Use the slider to change plot
    height, and browser window width to change plot width.<br>
    <br>
    Show / hide the <b>legend</b> and <b>0 |----</b> (zero) line,
    and&nbsp; plot using ggplot2 which is nicer, slower and draws the
    boxes slightly differently to base graphics. You can colour the
    boxes as individual modules by selecting <b>Module</b>, or as
    groups by their functional / descriptive category using <b>Title</b>,
    which may result in fewer boxes when multiple modules are grouped
    under one title.<br>
    <h3><a name="Modules-Genes_Modules_tab"></a> Modules-&gt;Genes
      (Modules) tab</h3>
    This expands on the <b>Selected Modules</b> tab&nbsp; to allow an
    inspection of the modules, one by one, looking at all the genes each
    module contains. It is essentially identical to <a
      href="#Modules-Genes_Probes_tab"><b>Modules-&gt;Genes</b></a> tab
    under <b>Explore By Spot</b>. Use the slider to change plot height,
    and browser window width to change plot width.<br>
    <br>
    Select a module from the menu to plot the summarised expression of
    spots mapping to its component genes automatically. <br>
    Show or hide the <b>legend</b> and <b>0 |---- (zero) line</b>,
    and&nbsp; use <b>ggplot2</b> instead of base graphics.&nbsp;
    ggplot2 uses a geom_boxplot with default values.<br>
    Genes listed for the module but absent from the dataset rows, are
    labelled "Missing" on the plot, which can be toggled on-off.
    <h3><a name="Modules:Series_Modules_tab"></a> Modules:Series
      (Modules) tab</h3>
    This is very similar to the <a href="#Modules:Series_Probes_tab"><b>Modules:Series






















        (Spots) tab</b></a> as it allows the expression of the selected
    modules to be plotted for all treatment ~ time combinations as a
    time course. Use the slider to change plot height, and browser
    window width to change plot width. The selected treatment~time
    column is indicated by <font color="#ff0000">▼▲<font
        color="#000000"> on the plot. </font></font>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    Click or drag over the points to identify the module,
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <font color="#ff0000"><font color="#000000">and you can copy the
        module names as a comma-separated list for pasting into keyword
        search boxes</font></font>. <br>
    <br>
    Select treatment ~ time combinations to include from the <b>Treatment</b>
    and <b>Days</b> menus. <br>
    <br>
    Select which modules to plot using one of the menus underneath:<br>
    <ul>
      <li> <b>Modules You Selected With Filters</b>&nbsp; lists the
        modules displayed in the <a
          href="#Selected_Modules_Modules_tab_"><b>Selected Modules tab</b></a>.
      </li>
      <li> <b>All Modules In Dataset</b> lists all the modules
        available in this dataset by their names.<br>
      </li>
      <li><b>All Module Titles In Dataset</b> menu lists all the
        available functional / descriptive module titles in this dataset
        - which group&nbsp; modules into categories.&nbsp;</li>
    </ul>
    <p>Once you have selected some modules or titles in the&nbsp; <b>All






















        Modules In Dataset</b> or <b>All Module Titles In Dataset</b>
      you can save the list by clicking the download button that
      appears. To repeat this selection (e.g. on another dataset) you
      must paste this comma-separated text into the <b>regex search box</b>
      on the <a href="#Select_modules_tab_"><b>Select Modules</b></a>
      tab. You cannot paste into these menus directly.</p>
    <p> Click <b>Plot</b> button to create the plot and data table, and
      after any changes to module/treatment/day selections or options.<br>
    </p>
    <b>Boxplot</b> plots a summary of the spot expressions in the
    modules<br>
    <b>Lines&nbsp; </b>plots the module mean expression and reveals <b>SEM</b>
    to include an SEM ribbon, plus <b>gridlines</b> and <b>points</b>.<br>
    <b>Split</b> plots each treatment in its own panel<br>
    <b>Legend</b> and<b> 0 |---- (zero)</b> line can be shown or hidden.
    <br>
    The bars can be coloured by individual modules when <b>Group By:</b>
    <b>Module</b> is selected, or by module <b>Title</b>, which may
    result in fewer boxes.<br>
    <br>
    <h2><a name="Transcriptomics_Lookup_tab_"></a>Transcriptomics Lookup
      tab<br>
    </h2>
    <h3><a name="Gene_Lookup_tab"></a>Gene Lookup tab</h3>
    <p>Enter a gene, spot or ProbeName and click <b>Lookup</b> to see a
      table of matching genes. Select Whole Word to avoid matching
      multiple stems (e.g. "CCL2" would match CCL2, CCL20, CCL21, etc.).
      The search uses <b>regex</b>, and <b>commas</b> to separate
      multiple search terms in the same way as selecting spots by
      keyword: <a href="#1._Regex_-_keyword_search_"><b>see that
          section for details</b></a>. Spaces will always be stripped
      from this search as Gene / Spot&nbsp; Names cannot have spaces. <br>
    </p>
    <p>You can paste the comma-separated lists downloaded by the various
      buttons on other tabs, to quickly describe the genes/spots.<br>
    </p>
    <p>Alternatively, leave the search box empty and click <b>Lookup</b>
      to list all 65k genes. The search boxes above / below the table
      can also be used to narrow the rows displayed.<br>
    </p>
    <h3><a name="Module_Lookup_tab"></a>Module Lookup tab</h3>
    <p>Enter a module name (e.g. <tt>DC.M4.1)</tt> or partial string
      (e.g. <tt>DC.M4</tt>) and click <b>Lookup</b> to see a table of
      matching modules. The search uses <b>regex</b>, and <b>commas</b>
      to separate multiple search terms in the same way as selecting
      spots by keyword: <a
href="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/help.html#1._Regex_-_keyword_search_"><b>see






















          that section for details</b></a>. <br>
      Spaces will always be stripped from this search as Module Names
      cannot have spaces.<br>
    </p>
    <p>Alternatively, leave the box empty and click <b>Lookup</b> to
      list all modules. The search boxes above / below the table can
      also be used. A list of modules with their descriptions and
      associated genes is shown. These gene lists can be copied from the
      table and directly pasted into the regex search boxes that accept
      Gene Names - spaces will be stripped for you.<br>
    </p>
    <h2><a name="Explore_Cytokines"></a>Explore Cytokines</h2>
    <p>The cytokines data is preloaded and immediately available. For
      details of the ABCD dataset <a href="#The_Datasets">see above</a>.
      This is a highly experimental tab and very slow to plot.<br>
    </p>
    <h3><a name="Plot_Cytokines"></a>Plot Cytokines</h3>
    <p> Use the menus to select some cytokines, vaccines and days. The
      order in which you enter the selections is respected in the plots,
      except for days on a continuous X axis. <br>
      Once you have selected some vaccines, days and cytokines you can
      click <b>Plot</b> to create plot, and to re-draw if you change
      any options. <b><br>
        This plot will take a long time if more than a few panels are
        plotted.</b><br>
    </p>
    <ul>
      <li><b>FI ~ Conc</b>. Use the radio buttons to switch between <b>fold





















          increase </b>(from day 0) or <b>actual concentration values</b>
        when plotting.FI may give some wide error bars if a few outliers
        have big changes.<br>
      </li>
      <li><b>Lines ~ Boxplot ~ Violin.</b><b> </b>Select which type of
        plot:<br>
        <b>Boxplot</b> is the usual box-whiskers (IQR), <b>Violin</b>
        is a frequency distribution. They both offer the option to <b>Crop





























          Y,</b> which zooms the plot to the IQR for the whole dataset -
        may be useful when extreme outliers squash the scale.<br>
        <b>Lines</b> plots the mean of the cytokine~treatment
        combination on each day. Further options appear when selected:<br>
      </li>
      <ul>
        <li><b>Points</b> adds a point to the lines.</li>
        <li><b>No Error ~ </b><b>Ribbon</b> ~ <b>Bars</b> : SEM
          plotted as ribbon or error bars</li>
      </ul>
      <li> Select<b> • |----&nbsp;</b> to plot a horizontal line at Y =
        1 or 0 (depends on whether log transformed). </li>
      <li><b>Exclude Zero Values</b>. The dataset already includes
        missing values (as NAs) where there was no data or "OOR"
        entered, or where a fold-increase could not be calculated (when
        day 0 was 0 or missing). It also contains many real zero values,
        where the concentration is entered as 0. To plot just the
        kinetics of the non-zero values, select this option and the data
        are filtered to remove zeros before means/medians/densities
        calculated.</li>
      <li><b>Log transforms</b>. Use the radio buttons to plot
        untransformed <b>Y</b>, or various log transforms. It is
        possible to efficiently mangle the data beyond all recognition
        or comprehension with these various options.<br>
        <b>ln1p</b> plots ln(1+Y) which is useful when there are zero
        values or Y &lt;&lt; 1.<br>
        For Boxplot and Violin plots the actual Y values are log
        transformed before any mean/medians are then calculated, so bear
        this in mind.<br>
        For Line plots the raw Y values are also transformed before
        calculating the mean - unless the <b>Means</b> checkbox is
        selected, in which case the mean is calculated from
        untransformed data then log transformed (equivalent to plotting
        on a log scale), and the SE is set to 0 (as log transformation
        of variance is tricky).<br>
        The raw data can always be downloaded from the data table to do
        your own transformations. <br>
        Zero values are removed <b>before any log transformation</b> if
        <b>Exclude Zero Y</b><b> Values</b> option is selected<br>
      </li>
      <li><b>Show N</b>. This prints the number of available data points
        per day at the bottom of the plot, and the actual number used at
        the top. The top number may be already be reduced by missing
        data in the dataset (NAs in the data), or when zero values have
        been omitted using <b>Exclude Zero Y</b><b> Values</b> option.
        The bottom number may also be reduced by data missing from the
        dataset (e.g. if one participant skipped a visit that reduces
        the group size for that day by 1). This plot is useful if <b>Exclude











          Zero Y Values</b> selected as it highlights how many
        participants have been excluded - or conversely how many do have
        useful data.<br>
      </li>
      <li><b>Fixed Y.</b> When selected the Y axis max-min range will be
        <b> fixed for each cytokine</b> across all treatments. Each
        cytokine will still have its own range, but the overall max -
        min will be imposed. This is relevant for placebo groups. If
        unselected, each panel has its own Y axis range.</li>
      <li><b>Vac↓ Cyt→</b> creates panels grouped by vaccine: cytokines
        wrap in the columns, and then a new block starts with each
        vaccine treatment. <br>
        <b>Vac→Cyt↓</b> does the reverse, with panels grouped by
        cytokine: vaccine treatments wrap in the columns, and then a new
        block starts with each cytokine.</li>
      <li><b>Grey</b> - when selected all plots are in greyscale.
        Otherwise I have pre-allocated <b>fixed colours to the
          cytokines</b> which do not change even when the number of
        cytokines plotted is changed. This allows easy cross-reference
        between plots with different combinations of cytokines.<br>
      </li>
      <li>A box to enter the <b>maximum</b><b> number of panels per row</b>,
        and slider to control the <b>height of the plot</b>, offer some
        control over the image. Widening or narrowing the window also
        changes the aspect ratio.<br>
      </li>
    </ul>
    The plot can be copied at low resolution by right-clicking, or a
    Hi-Res image downloaded by clicking the button. The data table can
    be downloaded by clicking the button.<br>
    <br>
    <h2><a name="Explore_Cells"></a>Explore Cells</h2>
    White blood cells data are not loaded automatically - click the
    button <b>Load Cells Data</b> to load an integrated dataset for
    trials ABC, D and D2.<br>
    Use the menus to select some cell types, vaccines and days. The
    order in which you enter the selections is respected in the plots,
    except for days on a continuous x-axis. <br>
    Once you have selected some vaccines, days and cell types you can
    click <b>Plot</b> to create plot, and to re-draw if you change any
    options. <b><br>
      This plot will take a long time if more than a few panels are
      plotted.</b><b><br>
    </b>
    <ul>
      <li><b>Lines ~ Boxplot:</b>Select which type of plot:</li>
      <ul>
        <ul>
          <li><b>Boxplot</b> is the usual box-whiskers (IQR).</li>
          <li><b>Lines</b> plots the mean of the cells type~treatment
            combination on each day. Further options appear when
            selected:<br>
          </li>
          <ul>
            <li><b>Points</b> adds a point to the lines. It is suggested
              that points are shown when <b>Split Vaccines</b> is
              unselected as otherwise it is not possible to distinguish
              the different vaccines plotted together.<br>
            </li>
            <li><b>SEM</b> : SEM plotted as ribbon</li>
            <li><b>Gridlines</b><br>
            </li>
          </ul>
        </ul>
      </ul>
      <li><b>Fold Change ~ Cell Count</b>. Use the radio buttons to
        switch between plotting <b>fold increase </b>(from day 0) or <b>actual














          cell values</b> when plotting.<br>
        When Cell Count Selected, the individual participant's cell
        counts are summarised on box plots, or the mean for the group
        plotted as a line chart.</li>
      <ul>
        <li> Select<b> • |----&nbsp;</b> to plot a horizontal line at Y
          = 1 when plotting <b>Fold Change</b></li>
      </ul>
      <li><b>Split Cells</b> creates a new panel to plot each cell type.
        If unselected then each cell type is plotted on one combined
        plot per vaccine: only really useful when Fold Change selected
        due to variations in normal ranges.</li>
      <li><b>Split Vaccines</b> creates a new row of plot panels to plot
        each cell type under a vaccine sub-heading when <b>Lines</b> is
        selected. If unselected then each cell type is plotted on one
        combined plot for all vaccines, or as a row if <b>Split Cells</b>
        is selected. <b>Split Vaccines</b> is always TRUE for box plots
        as it would be impossible to identify the vaccines if they were
        plotted on one plot.<br>
      </li>
      <li><b>Free Y</b>: When selected each plot panel has its own Y
        scale - useful only for looking at patterns. When unselected the
        Y scales are fixed by cell type across vaccines, but allowed to
        vary between cell types.</li>
      <li><b>Summary Legend</b>: produces one legend at the top of all
        plots, compared to ...</li>
      <li><b>Plot Legends: </b>where each plot panel has its own
        legend.</li>
      <li>A box to enter the <b>maximum</b><b> number of panels per row</b>,
        and slider to control the <b>height of the plot</b>, offer some
        control over the image. Widening or narrowing the window also
        changes the aspect ratio.</li>
    </ul>
    The plot can be copied at low resolution by right-clicking, or a
    Hi-Res image downloaded by clicking the button. The data table can
    be downloaded by clicking the button.<br>
    <br>
  </body>
</html>
