<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>Help</title>
  </head>
  <body>
    <h1> Introduction</h1>
    The purpose of this app is just to visualise the expression of genes
    as a time course after immunisation, either alone or when clustered
    into modules.<br>
    <br>
    The basic concept is to pick a particular time point - e.g. Day 7
    after immunisation with Boostrix, and then sort the probes or genes
    values present in the array as they are on that time point. So you
    may have the probes ordered in descending order by the value of that
    probe (fold-increase from baseline or mean expression) at that time
    in those people who got Boostrix. Having found the top or bottom
    probes, you can then select the probes or genes into a group of
    interest - e.g. the top 10 fold-increased genes on day 7 after
    Boostrix. This set of selected genes becomes the core of subsequent
    analysis. You can then identify which modules these genes belong to,
    and observe how other genes in those modules are responding over
    time, or how the whole module responds over time. You can treat
    these selected genes as a 'pseudo-module' and see how this module
    changes over time in comparison with the predefined modules in tmod.
    Other search options allow you to identify genes by their title or
    category, and then follow these genes and the modules that contain
    them over time. There is also an option to go straight to exploring
    by modules, in which case you just select modules and observe their
    response over time. Modules can be selected in the the same way as
    individual probes or genes - by identifying those most up- or
    down-regulated modules on a particular day after a particular
    treatment; or by searching for modules by name or category. But the
    basic concept is the same: how do the tmod modules respond after
    immunisation and how does this compare when we select individual
    genes? Are there other gene groupings that may be better? Do some
    modules have so many non-responding genes that they are not
    representative? Could we cluster new modules?<br>
    <br>
    This is beta software and is prone to bugs and crashes. Plots and
    data tables may take some time to appear - be patient! The more rows
    you select for plotting the slower it will be. ggplot2 is not good
    at providing progress updates. Some plots are drawn automatically
    and these offer the choice of using R base graphics (faster) or
    ggplot2 (slower but may be prettier).<br>
    <br>
    The software depends on the R package tmod '<cite>Feature Set
      Enrichment Analysis for Metabolomics and Transcriptomics</cite>' (<a
href="http://bioinfo.mpiib-berlin.mpg.de/tmod/tmod_paper_peerjpreprints_2016.pdf"
      target="_blank">description</a>), created by January Weiner. I
    have reverse engineered some aspects of tmod and any errors arising
    from this are mine. Consult the <a
      href="https://doi.org/10.7287/peerj.preprints.2420v1/supp-2"
      target="_blank">tmod documentation</a> for a description of the
    modules.The transcriptomics datasets are from the BIOVACSAFE project
    (<a href="http://www.biovacsafe.eu" target="_blank">www.biovacsafe.eu</a>),

    and were generated by the group of Stefan Kaufmann at MPIIB, Berlin
    and provided by Jeroen Maertzdorf and January Weiner. I have
    generated the probe averages per time point and any errors there are
    mine. The trials were conducted at the University of Gent by Geert
    Leroux-Roels, and at the University of Surrey by me. Source files
    for the shiny app are available at <a
      href="www.github.com/djmlewis/tmodexplorer" target="_blank">
      www.github.com/djmlewis/tmodexplorer</a>. The shiny app is online
    at <a href="djmlewis.shinyapps.io/tmodexplorer" target="_blank">djmlewis.shinyapps.io/tmodexplorer</a><br>
    <h2> Saving Plots &amp; Data Tables</h2>
    Plots can be copied or saved by right-clicking on the plot. The
    plotting parameters are not sophisticated, but generally the data
    used to create the plot are in the table below and can be saved and
    imported into R. Data tables can be downloaded by clicking the
    download button. Selected genes or probes can be downloaded as a
    text file in a comma-separated format suitable for pasting into the
    search box.<br>
    <h1> Instructions For Use</h1>
    <h2> Password</h2>
    When running on the shinyapps.io server the software requires a
    password. This is to protect the datasets. Please contact
    d.j.lewis@surrey.ac.uk to be given the password. Enter the password
    and click the button to unlock the menus. When running locally no
    password is required. <br>
    <h2> Loading Data</h2>
    You must load a dataset before you can do anything. The datasets are
    already uploaded to the server. Select a dataset. Click Load. All
    existing plots and data tables will be cleared. Note that some
    datasets are fold-increase from baseline, whereas others are actual
    expression values. The fold-increase are more useful. All data have
    been collated as a mean for each treatment group on each day. All
    expression values and fold-increases are log2. Once you click "Load"
    the name of the data set will appear below the tabs, and two more
    tabs will appear:<br>
    <ul>
      <li><a href="#Explore_By_Probe_tab">Explore By Probe</a></li>
      <li><a href="#Explore_By_Module_tab">Explore By Module</a></li>
    </ul>
    <h3><a name="Explore_By_Probe_tab"></a> Explore By Probe tab<br>
    </h3>
    Select this tab to explore the datasets using individual probes in
    the array. The idea is to choose a time point and a treatment, and
    then sort the probes based on their values (gene expression or
    fold-increase) in that column, and then additionally filter by
    keyword search or value. Note - if you include more that 200 rows
    then plots are slow to draw. Do not click "Plot" buttons or "Apply"
    again as the cycle just repeats!<br>
    <h4><a name="Select_tab_-_probes"></a> Select tab - probes<br>
    </h4>
    You must click the green <b>Apply Selections Below</b> button
    before any filters or selections are applied, and again after any
    changes. This is to speed-up calculations.<br>
    Pick a <b>column to sort</b> that has the treatment-time
    combination you want to explore. The "column" is the basic unit for
    tmodExplorer - it combines a treatment and a time point, e.g.
    "Boostrix_7" contains mean values on day 7 for participants who
    received Boostrix. All data sets are based on these columns.<br>
    <b>Sort descending</b> for highest values in top rows. Select <b>Gene
      Averages</b> to have expressions of individual probes that map to
    a gene averaged into one row, otherwise it will be one probe per row
    and genes may repeat. This is important if you restrict by "<b>Sorted
      Row Numbers</b>": e.g. if you select 10 rows and have <b>Gene
      Averages</b> selected you may include many more than 10 probes,
    but if <b>Gene Averages</b> is not selected you just get the top 10
    probes which may map to less than 10 genes. As the modules work in
    genes, it is probably better to select <b>Gene Averages</b>. Once
    you click <b>Apply Selections</b>, several other tabs will appear
    to allow visualisation of the genes and modules. The search terms
    you included appear below the tabs.<br>
    <h5> Filters</h5>
    Filters are applied one after another, left to right - first the
    keyword <b>regex</b> search, then the <b>value</b> restriction,
    finally which <b>rows</b> to include. Probably you will always use
    <b>Sorted Column Row Numbers</b> to restrict the number of sorted
    probes, possibly alone or in combination with either <b>regex</b>
    or <b>Sorted Column Values</b>. Never, ever search with no filters
    selected - it will return all 65k+ probes, <i>very slowly</i>.<br>
    <h6>Regex - keyword search<br>
    </h6>
    "Regex" is a keyword free-text search that uses regex, with some
    limitations. In tmodExplorer multiple terms can be entered separated
    by commas and are OR'd (search returns as many of those terms as it
    can find, treating each term separately). Do not include spaces
    (especially after the comma) or quotation marks unless you want them
    included in the search. Regex is "literal" so if you type "B cell"
    it is going to look for "B cell" - including the ". You can enter
    partial terms e.g. interferon finds interferon and interferons, as
    regex looks for those characters anywhere. Upper/lower case is
    ignored. Commas are interpreted as search term dividers and cannot
    be 'escaped'. Beware including <a href="https://regexr.com"
      target="_blank">special regex characters</a> by mistake. Select
    whether to search gene description or gene name. Regex actually
    ignores the selected column and any values - it just finds rows in
    the data set that match the search. However, those rows will be
    sorted in order of the value in the "Selected Column".<br>
    <h6>Sorted Column Values</h6>
    Max/min expression values will be pre-entered for you into the range
    boxes when selecting a column, using floor/ceil and so will
    deliberately exceed the actual values. Click Column to reset to
    max/min for selected column, or Data for max/min of whole dataset in
    all columns. Value restrictions apply only to values in the column
    selected for sort - so you will filter based on the values in the
    column you selected.<br>
    <h6>Sorted Column Row Numbers<br>
    </h6>
    It is probably a good idea to keep this selected, as over 200 rows
    is slow to calculate and plot. You can restrict rows to have top N
    or bottom N rows, or have a slice. If there are fewer rows available
    after the previous selections than you requested you will get a
    cryptic warning.<br>
    <h4><a name="Probes_tab"></a>Probes tab</h4>
    A data table showing the probes or genes (if averaged) that meet the
    filter settings is shown and can be downloaded. The values shown are
    those in the "Column" you selected. You may use the search boxes
    above and below the table to further restrict the rows for viewing
    in the table, but this has no effect on other tabs or plots. Click
    "Download Gene List" or "Download Probe List" to save a
    comma-separated text file that can be copy-pasted into the regex
    search box to quickly repeat the search. Note a description appears
    when probes are displayed individually, and disappears if "Gene
    Averages" is selected on the <b>Select</b> tab. This is because
    descriptions are specific for probes.<br>
    <h4> Probes:Series tab<br>
    </h4>
    This allows the values of the probes or gene averages selected by
    the filters to be plotted for over many treatment-time columns, to
    visualise a time course.<br>
    <h5> Columns menu <br>
    </h5>
    Click 'All' to enter all columns, or select one by one from pop-down
    menu and they appear in the box. Click 'None' to clear. The order of
    the columns as they appear in the box is respected for all boxes
    like this - so the x axis will retain the sequence you enter here,
    unless it is plotting as an XY scatter using numerical values. Use
    the cursor to move between entries to add or delete a column at that
    point. Select an entry and delete.<br>
    <h5>Options</h5>
    <b>Split</b>: the data are split by treatment using facet_wrap.
    Column names are in the format VVV_TTT where VVV is a treatment name
    and TTT is an integer value for time point. The app splits the
    column name by '_' and facet_wraps by VVV while plotting TTT on the
    x axis as a continuous variable.<br>
    <b>Points / Boxplot</b>: XY graph of individual probes / genes, or a
    boxplot summarising values when genes are averaged.<br>
    <b>Connect Points</b>: Appears if Points is selected and adds lines.<br>
    <b>Legend</b> and <b>Zero</b>: show/hide these features.<br>
    <b>Plot button</b>: click this to draw the plot, and again after
    changing any options.<br>
    <h4> Genes-&gt;Modules tab<br>
    </h4>
    This panel lists the selected probes or genes and shows which - if
    any - modules contain the gene. Some genes you selected may not be
    included in any modules.<br>
    <h4> Modules tab<br>
    </h4>
    A summary of the values of the individual probes contained in the
    modules associated with the selected probes or genes is plotted
    automatically. If you change any of the criteria on the <b>Select</b>
    tab this plot will re-draw. There are options to show or hide the <b>legend</b>
    and <b>zero</b> line, and to have ggplot2 graphics which are a bit
    nicer but slower. ggplot2 calculates the dimensions of the box
    slightly differently to base graphics. The genes selected by the
    filters can also be plotted as a "pseudo-module", for comparison
    with&nbsp; the tmod modules. The table gives the summary statistics
    and the number of probes mapped to the modules' genes (N).<br>
    <h4> Modules-&gt;Genes tab</h4>
    This expands the plot on the <b>Modules</b> tab&nbsp; to allow an
    inspection of the modules, one by one, looking at all the genes each
    module contains. The menu lists the modules associated with selected
    genes or probes that were summarised on the <b>Modules</b> tab.
    Select a module from the menu to see the expression of its component
    genes. As each gene may have many probes a summary boxplot of the
    expression values of the probes mapping to each of the the genes
    inside the selected module is plotted automatically. Note that a
    module may also contain genes that are not in the selection created
    by the <b>Select</b> tab. Genes within a module that are in your
    selection are denoted by the ︎► symbol next to the y axis, and in
    the data table. You may be surprised by how few genes in a module
    overlap with your selected genes. Options to show or hide the <b>legend</b>
    and <b>zero line</b> are available, and to use <b>ggplot2</b>
    instead of base graphics.&nbsp; ggplot2 uses a geom_boxplot with
    default values. Unlike the Selected Probes tab which shows the gene
    description in the table, the <b>Module-&gt;Genes</b> tab does not,
    as the descriptions refer to the <i>probes</i>, and here gene
    summaries are being shown.<br>
    <h4><a name="Modules:Series_Probes_tab"></a> Modules:Series (Probes)
      tab</h4>
    The <b>Modules</b> and <b>Modules-&gt;Genes</b> tabs only show the
    expression values of probes in the one selected treatment-time
    column. The <b>Modules:Series</b> tab allows the expression of the
    associated modules to be plotted for all treatment-time columns as a
    time course. To avoid delays when selecting options, nothing is
    plotted or re-plotted until the green 'Plot' button is clicked. Add
    treatment-time columns to the <b>Columns</b> menu. Add modules from
    the ones associated with your selected probes or genes to the
    <b> Modules</b> menu. Use <b>All</b> or <b>None</b> buttons to add
    all, or clear selections. Various plot options exist:<br>
    <b>Boxplot</b> summarises the expression values of probes within the
    module, while <b>Ribbon</b> plots a line graph of the mean
    expression of probes in the modules.<b> </b><b>Ribbon+SE</b>
    appears when <b>Ribbon</b> is selected allowing an SEM ribbon to be
    added.<br>
    <b>Split</b> - uses facet_wrap to separate treatments into different
    panels. <br>
    Options to show / hide the <b>legend</b> and <b>zero</b> line
    exist.<br>
    <h3><a name="Explore_By_Module_tab"></a> Explore By Module tab<br>
    </h3>
    The concept is similar to the <a href="#Explore_By_Probe_tab">Explore
      By Probe</a> tab. However, here a data table is created in which
    the rows are sorted by the mean or median value of the modules as
    expressed in the selected treatment-time column: e.g. the median
    expression of the probes on day 7 after Boostrix. It is important to
    note that tmod modules contain only a fraction of all the genes in
    the probeset, but this tab allows for an exploration of responses of
    the modules included in tmod.<br>
    <h4> Select tab - modules<br>
    </h4>
    Like <a href="#Select_tab_-_probes">Select tab</a> in <b>Explore
      By Probe</b> above, you first select a column containing the
    treatment-time combination for which you want to sort the modules.
    Then filter modules using the filters (left to right) as described
    in <a
href="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/help.html#Select_tab_-_probes">Select
      tab</a> for probes. In this case the sorting is based on the mean
    or median expression (which you can select) of the probes within the
    module. Sort descending has the highest expression values in top
    rows. See the <a
href="file:///Users/david/Dropbox/Hammersmith/PET/R_Analyses/Analyses/R/tmodExplorer/help.html#Select_tab_-_probes">Select
      tab</a> for description of how to use <b>regex</b>, <b>Sorted
      Column Values</b> and <b>Sorted Column Row Numbers</b>. Here it
    is the mean or median expression of the module being sorted. Regex
    searches either the <b>Title</b> (which is a description of the
    module, e.g. "plasma cells") or <b>Module</b> name (its designation
    e.g. DC.M4.11). You must click the <b>Apply Selections Below</b>
    button to perform the selection or after any changes to options.
    Clicking the button reveals more tabs to visualise the modules.<br>
    <h4> Selected Modules tab<br>
    </h4>
    Like the <a href="#Probes_tab"><b>Probes</b></a> tab this tab plots
    a boxplot summary of the probe expressions in the modules you
    selected. Options to show / hide the <b>legend</b> and <b>zero</b>
    line exist, and to plot using ggplot2 which is nicer, slower and
    draws the boxes slightly differently to base graphics. You can
    colour the boxes as individual modules by selecting <b>Module</b>,
    or as groups by their category using <b>Title.</b><br>
    <h4> Modules:Series (Modules) tab</h4>
    This is very similar to the <a href="#Modules:Series_Probes_tab">Modules:Series
      (Probes) tab</a>
    as it allows the expression of the selected modules to be plotted
    for all treatment-time columns as a time course. Select
    treatment-time columns to include from the <b>Columns</b> menu.
    Select which modules to plot using one of the menus underneath:<br>
    <b>Modules Selected By Filters</b> menu lists the modules displayed
    in the <b>Selected Modules tab</b>. <b><br>
      Titles In Dataset</b> menu lists all the available titles in this
    dataset - which group&nbsp; modules into functional categories.<br>
    <b>Modules In Dataset</b> just lists all the modules available in
    this dataset by their names.<br>
    Various plot options are available:<br>
    <b>Boxplot</b> plots a summary of the probe expressions in the
    modules, whereas <b>Ribbon&nbsp; </b>plots the module mean
    expression and reveals Ribbon+SE to include an SEM ribbon. <b>Split</b>
    plots each treatment in its own panel, and <b>legend</b> and <b>zero</b>
    line can be shown or hidden. <br>
    The bars can be coloured by individual modules when <b>Module</b>
    is selected, or by category <b>Title</b>.<br>
    Click <b>Plot</b> button to create the plot and data table, and
    after any changes to module/column selections or options.<br>
    <br>
    <h1> Versions</h1>
    1.2. Most features included and bugs trapped.<br>
    <br>
    © David JM Lewis www.djml.eu 2018. E&amp;OE<br>
    <br>
    <br>
  </body>
</html>
