<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>Help</title>
  </head>
  <body>
    <h1> Introduction</h1>
    The purpose of this app is just to visualise the expression of genes
    as a time course after immunisation, either alone or when clustered
    into modules.<br>
    <br>
    The basic concept is to pick a particular time point and vaccine - <i>e.g.





      Day 7 after immunisation with Boostrix</i> - and then sort the
    probes or genes values present in the array as they are on that time
    point in persons who received that vaccine. So you may have the
    probes ordered in descending order by their value (fold-increase
    from baseline or mean expression) at that time in those people who
    got Boostrix. Why do this? Well, because with over 60k probes you
    have to order them somehow. The grouping of vaccine - time point is
    actually a natural grouping and as you use the tool this becomes
    clear, if a bit counterintuitive when comparing between vaccines.<br>
    <br>
    Having ordered the top or bottom probes, you can then select the
    probes or genes into a group of interest - e.g. the top 10
    fold-increased genes on day 7 after Boostrix. <i><b>This set of
        selected genes becomes the core building-block of subsequent
        analysis</b></i>. You can identify how the expression of these
    genes changes over time; which modules these genes belong to;
    observe how the other genes in those modules are responding over
    time, or how the whole module responds over time. You can treat
    these selected genes as a 'pseudo-module' and see how this module
    changes over time in comparison with the predefined modules in tmod.
    Other search options allow you to identify genes by their title or
    category, and then follow these selected genes and the modules that
    contain them over time.<br>
    <br>
    There is also an option to go straight to exploring by modules, in
    which case you just select tmod modules and observe their response
    over time. Modules can be selected in the the same way as individual
    probes or genes - by identifying those most up- or down-regulated
    modules on a particular day after a particular treatment; or by
    searching for modules by name or category. But the basic concept is
    the same: how do the tmod modules respond after immunisation and how
    does this compare when we select individual genes? Are there other
    gene groupings that may be better? Do some modules have so many
    non-responding genes that they are not representative? Could we
    cluster new modules?<br>
    <br>
    <b>This is beta software and may experience bugs and crashes. </b>The





    software depends on the R package tmod '<cite>Feature Set Enrichment
      Analysis for Metabolomics and Transcriptomics</cite>' (<a
href="http://bioinfo.mpiib-berlin.mpg.de/tmod/tmod_paper_peerjpreprints_2016.pdf"
      target="_blank">description</a>), created by January Weiner. I
    have reverse engineered some aspects of tmod and any errors arising
    from this are mine. Consult the <a
      href="https://doi.org/10.7287/peerj.preprints.2420v1/supp-2"
      target="_blank">tmod documentation</a> or <a
      href="https://d3amtssd1tejdt.cloudfront.net/2016/2420/1/tmod.pdf"
      target="_blank">vignette</a> for a description of the modules.The
    transcriptomics datasets are from the BIOVACSAFE project (<a
      href="http://www.biovacsafe.eu" target="_blank">www.biovacsafe.eu</a>),















    and were generated by the group of Stefan Kaufmann at MPIIB, Berlin
    and provided by Jeroen Maertzdorf and January Weiner. I have
    generated the probe averages per time point and any errors there are
    mine. The trials were conducted at the University of Gent by Geert
    Leroux-Roels, and at the University of Surrey by me. Source files
    for the shiny app are available at <a
      href="http://www.github.com/djmlewis/tmodexplorer" target="_blank">
      www.github.com/djmlewis/tmodexplorer</a>. The shiny app is online
    at <a href="http://djmlewis.shinyapps.io/tmodexplorer"
      target="_blank">djmlewis.shinyapps.io/tmodexplorer</a><br>
    <h3>Plots</h3>
    <p><b><em>Plots may take some time to appear - be patient!</em></b>
      The more rows you select for plotting the slower it will be.
      ggplot2 is not good at providing progress updates - so it may look
      like nothing is happening. Some plots are drawn automatically and
      these offer the choice of using R base graphics (faster) or
      ggplot2 (slower but may be prettier). Other plots have a <b>Plot</b>
      button, and will only draw or update when that is clicked, to
      avoid slow refreshes when you alter several options. Plots have a
      slider control&nbsp;<img alt="" src="slider.png" width="48"
        height="20"> to change the height. Plot width always resizes
      automatically with the window width. A low resolution image of
      plots can be copied or saved by right-clicking on the plot, while
      clicking <b>⤓ Plot As PNG</b> button downloads a high resolution
      image. The plotting parameters are not sophisticated, but
      generally the data used to create the plot are in a table below
      and can be saved and imported into R for re-plotting. Some tables
      can be saved as PNG files to incorporate with plots as graphics.<br>
    </p>
    <h4>Identifying data on line plots by hovering over points<br>
    </h4>
    <p>When data are plotted as <b>Lines</b> it is sometimes difficult
      to identify the individual data series by colour. They can be
      identified by hovering the mouse cursor over a <b>point</b> (or
      where a point would be plotted if you are not displaying points).
      Hovering over line segments between points does nothing. A
      notification window will appear displaying the data for this
      point, which can be copied. It will disappear after 15 seconds, or
      be updated if a new point is selected. If lines are bunched then
      use the slider above the plot to increase the height.<br>
    </p>
    <p> </p>
    <h3>Data Tables<br>
    </h3>
    The entire data table can be searched using the box above,&nbsp; or
    individual columns searched using the boxes below each column.
    Tables can be downloaded as a csv file suitable for opening in Excel
    or R by clicking the <b>⤓ Table</b> button. A PNG image of the
    datatable can also be downloaded in some cases by clicking the
    button <b>⤓ Table as PNG</b>. This can be pasted onto graphics. It
    is very tricky to calculate the width correctly using grid.table, so
    the image may be clipped or extra wide, but a transparent background
    will be generated.<br>
    <br>
    <h3>Saving Gene, Probe and Module Lists<br>
    </h3>
    When you have selected genes, probes or modules using some filters,
    they can be downloaded as a text file in a comma-separated format
    suitable for pasting into the 'regex' search boxes.<br>
    <h1>The Datasets</h1>
    <p>The preloaded datasets come in two versions: <b>fold-increase in
        the log2 expression values</b> (by subtraction of the log2
      values), <i>always from day 0</i> <i>baseline</i> (in Surrey day
      -1 is calculated as a time point preceding day 0); and <b>raw
        log2 expression values</b>. Day 0 at 8AM was selected for Surrey
      as baseline as participants entered the unit on day -1. The data
      were calculated by initially grouping the raw data for each
      person-each probe-each timepoint, into treatment and timepoint
      pools. Then row averages for each probe within the treatment-time
      pool were calculated. This gives the basic raw expression data
      which is the group mean for each vaccination-day for each probe. <br>
    </p>
    <p>One group of datasets, based on <i><b>limma-normalised</b></i>
      expression values, consists of the <b> protocol-defined grouping
        of the vaccines and associated placebo group</b> (which may have
      only ~6 participants in Surrey):<br>
    </p>
    <ul>
      <li>A Surrey Stamaril-Varilrix-Placebo</li>
      <li>C Surrey Aggrippal-Fluad-Placebo</li>
      <li>D Gent Fluad-Placebo</li>
      <li>D2 Gent Boostrix-Placebo</li>
    </ul>
    <p>An exception is <b>Surrey Engerix B</b>. Here the participants
      were naive and received a priming vaccination alongside a placebo
      group: this is indicated as ENGERIXB and PLACEBO in the same way
      as the others above. The same participants then had a second
      immunisation (first boost) without any systems biology. They then
      came for a third immunisation (second boost) on day 167. This is
      represented in the dataset as EngerixB3 and PLACEBO3. The time
      points for ENGERIXB3/PLACEBO3 have been adjusted so that 167
      becomes -1, and 168 day 0. <i>Fold-increase is then calculated
        from this adjusted day 0</i>, as the response to the third
      immunisation. Thus the ENGERIXB and PLACEBO may be pooled with the
      other Surrey trials, but the ENGERIXB3 and especially the PLACEBO3
      should be treated with caution as they are the same persons
      repeated. This dataset therefore has the pattern:<br>
    </p>
    <ul>
      <li>B Surrey EngerixB-Placebo-EngerixB3-Placebo3</li>
    </ul>
    I have also included a <b>combined Surrey dataset</b> in which the
    vaccines in A, B and C are all together, based on <i><b>agilent</b></i><i><b>-normalised</b></i>
    expression values. In this case the PLACEBO group contains all ~21
    participants pooled from ABC - <i>without</i> the PLACEBO3 group as
    described for <i>B Surrey Engerix B</i> above. PLACEBO3 (the
    subjects who repeated) are presented as a separate group. This is a
    large dataset and so can be slow, but allows very convenient
    comparison between vaccines. This dataset therefore has the pattern:
    <ul>
      <li>ABC Surrey Agilent</li>
    </ul>
    <ul>
    </ul>
    <h1> </h1>
    <h1> Instructions For Use</h1>
    <h2> Password</h2>
    When running on the shinyapps.io server the software requires a
    password. This is to protect the datasets. Please contact
    d.j.lewis@surrey.ac.uk to be given the password. Enter the password
    and click the button to unlock the menus. When running locally no
    password is required. <br>
    <h2> Loading Data</h2>
    You must load a dataset before you can do anything. The datasets are
    already uploaded to the server. Select a dataset. Click Load. All
    existing plots and data tables will be cleared. Note that some
    datasets are fold-increase from baseline, whereas others are actual
    expression values. The fold-increase are more useful. All data have
    been collated as a mean for each treatment group on each day. All
    expression values and fold-increases are log2. Once you click "Load"
    the name of the data set will appear below the tabs, and two more
    tabs will appear:<br>
    <ul>
      <li>Explore By Probe</li>
      <li>Explore By Module</li>
    </ul>
    <h2><a name="Explore_By_Probe_tab"></a> Explore By Probe tab<br>
    </h2>
    Select this tab to explore the datasets using individual probes in
    the array. The idea is to choose a time point and a treatment, and
    then sort the probes based on their values (gene expression or
    fold-increase) in that column, and then additionally filter by
    keyword search or value. Note - if you include more that 200 rows
    then plots are slow to draw. Do not click "Plot" buttons or "Apply"
    again as the cycle just repeats!<br>
    <h3><a name="Select_tab_-_probes"></a> Select tab - probes<br>
    </h3>
    You must click the green <b>Apply Selections</b> button before any
    filters or selections are applied, and again after any changes. This
    is to speed-up calculations.<br>
    Pick a <b>column to sort</b> that has the treatment-time
    combination you want to explore. The "column" is the basic unit for
    tmodExplorer - it combines a treatment and a time point, e.g.
    "Boostrix_7" contains mean values on day 7 for participants who
    received Boostrix. All data sets are based on these columns. Time
    series plots will indicate the sorted column with <font
      color="#ff0000">▼▲</font>.<br>
    <b>Sort descending</b> for highest values in top rows. Select <b>Gene














      Averages</b> to have expressions of individual probes that map to
    a gene averaged into one row, otherwise it will be one probe per row
    and genes may repeat. This is important if you restrict by "<b>Sorted














      Row Numbers</b>": e.g. if you select 10 rows and have <b>Gene
      Averages</b> selected you may include many more than 10 probes,
    but if <b>Gene Averages</b> is not selected you just get the top 10
    probes which may map to less than 10 genes. As the modules work in
    genes, it is probably better to select <b>Gene Averages</b>, unless
    you specifically want to follow distinct probes. A disadvantage of
    gene averages is that you lose the gene descriptions, which are
    probe-specific. Toggling between averaged and probe-alone settings
    lets you see what these genes are, alternatively use the <b>Gene
      Lookup tab</b> to lookup genes by their names and see the probe
    descriptions.<br>
    <br>
    Once you click <b>Apply Selections</b>, several other tabs will
    appear to allow visualisation of the genes and modules. The search
    terms you included appear below the tabs.<br>
    <h4> Filters</h4>
    Filters are applied one after another, left to right - first the
    keyword <b>regex</b> search, then the <b>value</b> restriction,
    finally which <b>rows</b> to include. Probably you will always use
    <b>Sorted Column Row Numbers</b> to restrict the number of sorted
    probes, possibly alone or in combination with either <b>regex</b>
    or <b>Sorted Column Values</b>. Never, ever try to search with no
    filters selected - it will return all 65k+ probes, <i>very slowly,
      and I have set a block on this.</i><br>
    <h5>Regex - keyword search<br>
    </h5>
    "Regex" is a keyword, free-text search that uses regex, with some
    limitations. In tmodExplorer multiple terms can be entered into the
    box separated by commas and are OR'd (search returns as many of
    those terms as it can find, treating each term separately). Do not
    include spaces (especially after the comma) or quotation marks
    unless you want them included in the search. Spaces will be
    stripped-out automatically for you in <i>Gene</i> and <i>Probe</i>
    name searches, so you can easily paste in lists copied from tables.
    For <i>Description</i> searches the spaces are left in so beware:
    regex is "literal" so if you type <tt>"B cell"</tt> it is going to
    look for <tt>"B cell"</tt> - including the <tt>""</tt>. You can
    enter partial terms e.g. interferon finds interferon and
    interferons, as regex looks for those characters anywhere.
    Upper/lower case is ignored. Commas are interpreted as search term
    dividers and cannot be 'escaped'. Beware including <a
      href="https://regexr.com" target="_blank">special regex characters</a>
    by mistake. Select whether to search gene description or gene name,
    or probe ID. Probe is least useful as they are not very standardised
    between datasets.<br>
    <br>
    Regex search actually ignores the selected column and any values -
    it just finds rows in the data set that match the search. However,
    those rows will be sorted in order of the value in the "Selected
    Column", and filtered on these values. You may want to switch off
    the <b>Sorted Column Row Numbers</b> when using regex as it may
    reduce the number of free-text hits you would have found. However,
    beware returning hundreds of rows as the app will freeze for a long
    time. Consider setting some <b>Sorted Column Values</b> limits. You
    can also <b>paste saved search lists</b> into the regex box that
    you downloaded from the <b>Selected Probes tab</b>, or copied from
    one of the <b>Lookup</b> tables - be sure to match Gene and Probe
    Name. <b>This allows fairly easy repetition of searches across
      datasets</b>.&nbsp;
    <h5>Sorted Column Values</h5>
    Max/min expression values will be pre-entered for you into the range
    boxes when selecting a column, using floor() and ceil() and so will
    deliberately exceed the actual values. Click <b>Column</b> to reset
    to max/min for selected column, or <b>Data</b> for max/min of whole
    dataset in all columns. Value restrictions apply only to values in
    the column selected for sort - <b>so you will filter based on the
      values in the column you selected</b>.<br>
    <h5>Sorted Column Row Numbers<br>
    </h5>
    It is probably a good idea to keep this selected, as over 200 rows
    is slow to calculate and plot. You can restrict rows to have top N
    or bottom N rows, or have a slice. If there are fewer rows available
    after the previous selections than you requested you will get a
    cryptic warning. Row restrictions apply&nbsp; to the sorted values
    in the column selected for sort - <b>so you will filter based on
      the values in the column you selected</b>.
    <h3><a name="Probes_tab"></a>Selected Probes tab</h3>
    A data table showing the probes or genes (if averaged) that meet the
    filter settings is shown and can be downloaded. The values shown are
    those in the "Column" you selected. You may use the search boxes
    above and below the table to further restrict the rows for viewing
    in the table, but this has no effect on other tabs or plots. Click
    "⤓ Gene List" or "⤓ Probe List" to save a comma-separated text file
    that can be copy-pasted into the regex search box to quickly repeat
    the search. Note a description appears when probes are displayed
    individually, and disappears if "Gene Averages" is selected on the <b>Select</b>
    tab. This is because descriptions are specific for probes.<br>
    <h3> Probes:Series tab<br>
    </h3>
    This allows the values of the probes or gene averages selected by
    the filters to be plotted for over many treatment-time columns, to
    visualise a time course.<br>
    <h4> Columns menu <br>
    </h4>
    Click 'All' to enter all columns, or select one by one from pop-down
    menu and they appear in the box. Click 'None' to clear. The order of
    the columns as they appear in the box is respected for all boxes
    like this - so the x axis will retain the sequence you enter here,
    unless it is plotting as an XY scatter using numerical values. Use
    the cursor to move between entries to add or delete a column at that
    point. Select an entry and delete.<br>
    <h4>Options</h4>
    <b>Split</b>: the data are split by treatment using facet_wrap.
    Column names are in the format VVV_TTT where VVV is a treatment name
    and TTT is an integer value for time point. The app splits the
    column name by '_' and facet_wraps by VVV while plotting TTT on the
    x axis as a continuous variable.<br>
    <b>Lines / Boxplot</b>: XY graph of individual probes / genes, or a
    boxplot summarising values when genes are averaged.<br>
    <b>Points</b>, <b>Gridlines</b>: show/hide.<br>
    <b>Legend</b> and <b>Zero</b>: show/hide these features.<br>
    <b>Plot button</b>: click this to draw the plot, and again after
    changing any options.<br>
    <h3> Genes-&gt;Modules tab<br>
    </h3>
    This panel lists the selected probes or genes and shows which - if
    any - modules contain the gene. Some genes you selected may not be
    included in any modules.<br>
    <h3> Modules tab<br>
    </h3>
    A summary of the values of the individual probes contained in the
    modules associated with the selected probes or genes is plotted
    automatically. If you change any of the criteria on the <b>Select</b>
    tab this plot will re-draw. There are options to show or hide the <b>legend</b>
    and <b>zero</b> line, and to have ggplot2 graphics which are a bit
    nicer but slower. ggplot2 calculates the dimensions of the box
    slightly differently to base graphics. The genes selected by the
    filters can also be plotted as a "pseudo-module", for comparison
    with&nbsp; the tmod modules. The table gives the summary statistics
    and the number of probes mapped to the modules' genes (N).<br>
    <h3> Modules-&gt;Genes tab</h3>
    This expands the plot on the <b>Modules</b> tab&nbsp; to allow an
    inspection of the modules, one by one, looking at all the genes each
    module contains. The menu lists the modules associated with selected
    genes or probes that were summarised on the <b>Modules</b> tab.
    Select a module from the menu to see the expression of its component
    genes. As each gene may have many probes a summary boxplot of the
    expression values of the probes mapping to each of the the genes
    inside the selected module is plotted automatically. Note that a
    module may also contain genes that are not in the selection created
    by the <b>Select</b> tab. Genes within a module that are in your
    selection are denoted by the ︎► symbol next to the y axis, and in
    the data table. You may be surprised by how few genes in a module
    overlap with your selected genes. Options to show or hide the <b>legend</b>
    and <b>zero line</b> are available, and to use <b>ggplot2</b>
    instead of base graphics.&nbsp; ggplot2 uses a geom_boxplot with
    default values. Unlike the Selected Probes tab which shows the gene
    description in the table, the <b>Module-&gt;Genes</b> tab does not,
    as the descriptions refer to the <i>probes</i>, and here gene
    summaries are being shown.<br>
    <h3><a name="Modules:Series_Probes_tab"></a> Modules:Series (Probes)
      tab</h3>
    The <b>Modules</b> and <b>Modules-&gt;Genes</b> tabs only show the
    expression values of probes in the one selected treatment-time
    column. The <b>Modules:Series</b> tab allows the expression of the
    associated modules to be plotted for all treatment-time columns as a
    time course. To avoid delays when selecting options, nothing is
    plotted or re-plotted until the green 'Plot' button is clicked. Add
    treatment-time columns to the <b>Columns</b> menu. Add modules from
    the ones associated with your selected probes or genes to the <b>
      Modules</b> menu. Use <b>All</b> or <b>None</b> buttons to add
    all, or clear selections. Various plot options exist:<br>
    <b>Boxplot</b> summarises the expression values of probes within the
    module, while <b>Lines</b> plots a line graph of the mean
    expression of probes in the modules.<b> SEM</b> appears when <b>Lines</b>
    is selected allowing an SEM ribbon to be added.<br>
    <b>Split</b> - uses facet_wrap to separate treatments into different
    panels. <br>
    Options to show / hide the <b>legend</b> and <b>zero</b> line
    exist.<br>
    <h3>Gene Lookup tab</h3>
    <p>Enter a gene name or partial string and click <b>Lookup</b> to
      see a table of matching genes. The search uses <b>regex</b>, so
      see the entries above about filtering by keyword for issues with
      regex. Spaces will always be stripped from this search as Gene
      Names cannot have spaces. Multiple gene names can be entered
      separated by commas. Alternatively, leave the box empty and click
      <b>Lookup</b> to list all 65k genes. A list of probes with their
      descriptions is shown. The search boxes above / below the table
      can also be used.<br>
    </p>
    <h3> </h3>
    <h2><a name="Explore_By_Module_tab"></a> Explore By Module tab<br>
    </h2>
    The concept is similar to the <b>Explore By Probe</b> tab. However,
    here a data table is created in which the rows are sorted by the
    mean or median value of the modules as expressed in the selected
    treatment-time column: e.g. the median expression of the probes on
    day 7 after Boostrix. It is important to note that tmod modules
    contain only a fraction of all the genes in the probeset, but this
    tab allows for an exploration of responses of the modules included
    in tmod.<br>
    <h3> Select tab - modules<br>
    </h3>
    Like Select tab in <b>Explore By Probe</b> above, you first select
    a column containing the treatment-time combination for which you
    want to sort the modules. Then filter modules using the filters
    (left to right) as described in Select tab for probes. In this case
    the sorting is based on the mean or median expression (which you can
    select) of the probes within the module. Sort descending has the
    highest expression values in top rows. See the Select tab for
    description of how to use <b>regex</b>, <b>Sorted Column Values</b>
    and <b>Sorted Column Row Numbers</b>. Here it is the mean or median
    expression of the module being sorted. Time series plots will
    indicate the sorted column with <font color="#ff0000">▼▲</font>.<br>
    <br>
    <b>Regex</b> searches either the <b>Title</b> (which is a
    description of the module, e.g. "plasma cells") or <b>Module</b>
    name (its designation e.g. DC.M4.11). Spaces will automatically be
    stripped from Module searches but retained for Title searches. <br>
    <br>
    You may want to switch off the <b>Sorted Column Row Numbers</b>
    when using regex as it may reduce the number of hits you would have
    found. However, beware returning 600 modules as the app will freeze
    for a long time. Consider setting some <b>Sorted Column Values</b>
    limits instead - this will be based on values in the selected sort
    column. You can also <b>paste saved search lists</b> into the regex
    box that you downloaded from the <b>Selected Modules tab</b> - be
    sure to match Title and Module Name. This allows fairly easy
    repetition of searches across datasets.<br>
    <br>
    You must click the <b>Apply Selections</b> button to perform the
    selection or after any changes to options. Clicking the button
    reveals more tabs to visualise the modules.<br>
    <h3> Selected Modules tab<br>
    </h3>
    Like the <b>Probes</b> tab this tab plots a boxplot summary of the
    probe expressions in the modules you selected. Options to show /
    hide the <b>legend</b> and <b>zero</b> line exist, and to plot
    using ggplot2 which is nicer, slower and draws the boxes slightly
    differently to base graphics. You can colour the boxes as individual
    modules by selecting <b>Module</b>, or as groups by their category
    using <b>Title.</b><br>
    <h3> Modules:Series (Modules) tab</h3>
    This is very similar to the <b>Modules:Series (Probes) tab</b> as
    it allows the expression of the selected modules to be plotted for
    all treatment-time columns as a time course. Select treatment-time
    columns to include from the <b>Columns</b> menu. Select which
    modules to plot using one of the menus underneath:<br>
    <b>Modules Selected By Filters</b> menu lists the modules displayed
    in the <b>Selected Modules tab</b>. <b><br>
      Titles In Dataset</b> menu lists all the available titles in this
    dataset - which group&nbsp; modules into functional categories. You
    can save the selection by clicking the download button - to repeat
    this selection (e.g. on another dataset) you must paste the
    comma-separated text into the <b>regex search box</b> on the select
    tab. You cannot paste into this menu box.<br>
    <b>Modules In Dataset</b> just lists all the modules available in
    this dataset by their names. You can save the selection by clicking
    the download button, but to repeat this selection (e.g. on another
    dataset) you must paste the comma-separated text into the <b>regex
      search box</b> on the select tab. You cannot paste into this menu
    box.<br>
    Various plot options are available:<br>
    <b>Boxplot</b> plots a summary of the probe expressions in the
    modules, whereas <b>Lines&nbsp; </b>plots the module mean
    expression and reveals <b>SEM</b><b></b> to include an SEM ribbon.
    <b>Split</b> plots each treatment in its own panel, and <b>legend</b>
    and <b>zero</b> line can be shown or hidden. <br>
    The bars can be coloured by individual modules when <b>Module</b>
    is selected, or by category <b>Title</b>.<br>
    Click <b>Plot</b> button to create the plot and data table, and
    after any changes to module/column selections or options.<br>
    <h3>Module Lookup tab</h3>
    <p>Enter a module name (e.g. <tt>DC.M4.1)</tt> or partial string
      (e.g. <tt>DC.M4</tt>) and click <b>Lookup</b> to see a table of
      matching modules. The search uses <b>regex</b>, so see the
      entries above about filtering by keyword for issues with regex.
      Spaces will always be stripped from this search as Module Names
      cannot have spaces. Multiple module names can be entered separated
      by commas. Alternatively, leave the box empty and click <b>Lookup</b>
      to list all modules. The search boxes above / below the table can
      also be used. A list of modules with their descriptions and
      associated genes is shown. These gene lists can be copied and
      directly pasted into the regex search boxes that accept Gene
      Names, spaces will be stripped for you. </p>
    <h2>Acknowledgements</h2>
    tmodExplorer is written in R: R Core Team (2017). R: A language and
    environment for statistical computing. R Foundation for Statistical
    Computing, Vienna,<br>
    &nbsp; Austria. URL <a href="https://www.R-project.org/."
      target="_blank">https://www.R-project.org/. </a>tmodExplorer uses
    tidyverse packages (<a href="https://www.tidyverse.org"
      target="_blank">https://www.tidyverse.org</a>) extensively
    especially dplyr, purrr, readr, ggplot2 and tibble.<br>
    <h2> Versions</h2>
    1.2beta. Most features included and bugs trapped.<br>
    <br>
    University Of Surrey © 2018. E&amp;OE David JM Lewis&nbsp;
    🏴󠁧󠁢󠁷󠁬󠁳󠁿<br>
    <br>
    <br>
  </body>
</html>
