<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>QuickStart</title>
  </head>
  <body>
    <ol>
    </ol>
    <h2>Quick Start On Exploring Transcriptomics Datasets<br>
    </h2>
    <ol>
      <li><b>Load transcriptomics</b> <b>tab</b>. Choose a dataset to
        explore from the menu and click <b>Load Dataset.</b> <br>
        The dataset appears showing some annotation data in the first
        columns and then the log2 normalised gene expression values.
        Datasets have expression values of individual spots as rows
        (mapping to ProbeNames (nucleic acid sequences) and genes - one
        ProbeName may map to several spots, and one Gene may map to
        several ProbeNames). Each column is the mean expression for one
        vaccine~day combinations, where participants have been grouped
        by treatment (e.g. BOOSTRIX_1 column is group mean for Boostrix
        vaccinated participants on Day 1). Both Fold Change (from Day 0)
        and Raw Expression value datasets are available.<br>
      </li>
      <li><b>Explore By Gene</b> tab: Select this and the following
        options appear:<br>
      </li>
      <ol>
        <li><b>Select Genes</b>: Filter the rows of spot expression
          values by some criteria. If you click <b>Apply Filters</b>
          with the default setting the 10 rows with spots having the
          highest value in the column matching the Vaccine~Day
          combination selected in the menus will be selected and all
          other rows filtered-out. <br>
          Select <b>Gene Averages</b> to select top 10 genes based on
          the mean gene expression values (calculated from all spots
          mapping to probes that map to that gene). instead of
          spots.&nbsp; As returning more than 100 rows from a filter
          will take a long time to display, at the top right is a box to
          enter a maximum number of rows to return whatever the filter
          settings - the default is 100.<br>
          Other filter options include <b>keyword</b> to select rows
          using a text-based filter on the annotation columns; and by <b>value</b>
          of the rows in the selected column or across all columns. See
          the detailed instructions for further information.<br>
          <br>
          <i> Once you have clicked "Apply Filters" the following tabs
            now appear:</i><br>
        </li>
        <ol>
        </ol>
        <li><b>Selected Genes</b>&nbsp; lists the selected rows (spots
          or genes) that meet the filter criteria.</li>
        <ol>
        </ol>
        <li><b>Genes:Series</b>: plot the time course of selected rows
          values. Select some vaccines and days to plot, and the spots /
          genes to plot. Click <b>Plot</b> button each time you change
          options.</li>
        <ol>
        </ol>
        <li><b>Genes-&gt;Modules</b> lists genes matching selected rows,
          and identifies which modules contain these genes (if any).
          Filtered spots/genes that do not match any module are not
          listed - pay attention to the Rank column as it assists with
          understanding which rows are missing because they do not match
          any modules.<br>
        </li>
        <ol>
        </ol>
        <li><b>Modules</b> box plot of the all the row values which map
          to genes contained within the selected modules. Include the
          filters-selected genes as a separate "pseudo-module". Plots
          automatically after every options change. <b>Note:</b> the
          values are from the treatment-time combination selected in the
          menus in <b>Select Genes</b> - to see a different time or
          treatment you must change the filters and re-apply.<br>
        </li>
        <ol>
        </ol>
        <li><b>Module-&gt;Genes</b>&nbsp; For every selected module it
          lists the genes contained within that module. Select a module
          from the menu to plot a box plot summarising the row values
          for each gene (some genes may have only one spot, others many
          probes/spots). Genes that also passed the filters indicated by
          ▶︎. Plots automatically. Some modules may have very few
          filtered genes, or a wide spread of gene expressions. Reflects
          how the clustering of genes within a module may not reflect
          the clustering of genes in response to the vaccines.</li>
        <ol>
        </ol>
        <li><b>Module:Series</b> plot the time course of row values
          within each module. Select some vaccines and days, and some
          modules from those identified by the genes filter. Click Plot
          button to create or update plot. Include the selected genes as
          a separate "pseudo-module".</li>
        <ol>
        </ol>
      </ol>
      <li><b>Explore By Module</b> tab:<br>
        This works on the same principles as <b>Explore By Gene</b>
        except you filter modules by keyword or based on the mean or
        median value of all the spots mapping to that module. </li>
      <ol>
        <li><b>Select Module</b> filter modules by mean or median spot
          values, or keyword. Click <b>Apply Selection</b>.<br>
          <i> Once you have clicked "Apply Filters" the following tabs
            now appear:</i><br>
        </li>
        <ol>
        </ol>
        <li><b>Selected Modules</b> - box plot of the all the row values
          which map to genes contained within the selected modules.
          Plots automatically.</li>
        <ol>
        </ol>
        <li><b>Module-&gt;Gene</b>s For every selected module it lists
          the genes contained within that module and then plots a box
          plot summarising the row values for each gene (some genes may
          have only one spot, others many probes/spots). Select a module
          from the menu. Plots automatically. Some modules may have very
          few upregulated genes, or a wide spread of gene expressions.
          Reflects how the clustering of genes within a module may not
          reflect the clustering of genes in response to the vaccines. <b>Note:</b>
          the values are from the treatment-time combination selected in
          the menus in <b>Select Module</b> - to see a different time
          or treatment you must change the filters and re-apply. </li>
        <ol>
        </ol>
        <li><b>Module:Series</b> plot the time course of row values
          within each module. Select some vaccines and days, and some
          modules. Click <b>Plot</b> button each time you change
          options. Select modules to plot from the options: modules
          selected in <b>Select Modules</b> tab; <b>all modules
            available</b> in database, all modules available in database
          clustered under their functional <b>Title</b> categories.<br>
          <br>
          <br>
        </li>
        <ol>
        </ol>
      </ol>
      <li>For help on the various options presented on the screens and
        for <b>Lookup</b>, <b>Cells</b> and <b>Cytokines</b>: see <b>Detailed


          Instructions For Use</b> in <b>ReadMe</b>.<br>
      </li>
    </ol>
  </body>
</html>
