<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>QuickStart</title>
  </head>
  <body>
    <ol>
    </ol>
    <h2>Quick Start On Exploring Transcriptomics Datasets<br>
    </h2>
    <ol>
      <li>Enter the <b>password</b> if using the app on shinyapps.io
        server. Click Enter</li>
      <li><b>Load transcriptomics</b> tab. Choose a dataset to explore
        from the menu and click <b>Load Dataset.</b> Dataset has
        individual spots in rows (mapping to ProbeNames (nucleic acid
        sequences) and genes - one ProbeName may map to several spots,
        and one Gene may map to several ProbeNames) and vaccine~day
        combinations in columns (group mean). Fold Increase and Raw
        Expression values are available.<br>
      </li>
      <li><b>Explore By Spot</b> tab: <br>
      </li>
      <ol>
        <li><b>Select Spots</b> tab: select a vaccine and a day to sort
          the spots by. If you click <b>Apply Filters</b> with the
          default setting the top 10 spots for selected vaccine on
          selected day are selected. Select <b>Gene Averages</b> to
          select top 10 genes instead of spots. Apply other search
          filters to select genes / spots. <br>
          The following tabs now appear:<br>
        </li>
        <ol>
          <li><b>Selected Spots</b> tab lists the selected spots / genes
            with descriptions for spots.</li>
          <li><b>Spots:Series</b> tab: Select some vaccines from
            Treatment menu, and Days. Click <b>Plot</b> button.
            Kinetics of selected spots appears. Click <b>Plot</b>
            button each time you change options.</li>
          <li><b>Genes-&gt;Modules</b> tab lists the genes matching
            selected spots / genes, and which mapped modules contain
            these genes (if any).</li>
          <li><b>Modules</b> tab summarises the spot values within each
            mapped module that map the selected spots / genes. Include
            the selected genes as a "pseudo-module". Plots
            automatically.<br>
          </li>
          <li><b>Module-&gt;Genes</b> tab summarises the spot values for
            genes within each mapped module, one by one. Select a module
            from the menu and the spot values summarised as boxplots for
            each gene. Plots automatically. Some modules may have very
            few upregulated genes, or a wide spread of gene expressions.
            Reflects how the clustering of genes within a module may not
            reflect the clustering of genes in response to vaccines.</li>
          <li><b>Module:Series</b> tab allows the kinetics of the spot
            values within each module to be plotted over time. Select
            some vaccines from Treatment menu, and Days. Click Plot
            button to create plot. Click <b>Plot</b> button each time
            you change options. Include the selected genes as a
            "pseudo-module".</li>
        </ol>
      </ol>
      <li><b>Explore By Module</b> tab: <br>
      </li>
      <ol>
        <li><b>Select Module</b> tab: select a vaccine ~ day combination
          by which to sort the tmod modules mean or median spot values.
          Click <b>Apply Selection</b>. Top 10 modules for selected
          vaccine on selected day are selected. Use other search filters
          to select modules.<br>
          The following tabs now appear:<br>
        </li>
        <ol>
          <li><b>Selected Modules</b> tab summarises the spot values
            within each selected module. Plots automatically.</li>
          <li><b>Module-&gt;Gene</b>s tab summarises the spot values for
            genes within each selected module, one by one. Select a
            module from the menu and the spot values summarised as
            boxplots for each gene. Plots automatically. Some modules
            may have very few upregulated genes, or a wide spread of
            gene expressions. Reflects how the clustering of genes
            within a module may not reflect the clustering of genes in
            response to vaccines.<br>
          </li>
          <li><b>Module:Series</b> tab allows the kinetics of the spot
            values within each selected module to be plotted over time.
            Select some vaccines from Treatment menu, and Days. Click
            Plot button to create plot. Click <b>Plot</b> button each
            time you change options. Select modules to plot from the
            options: modules selected in <b>Select Modules</b> tab; <b>all





              modules available</b> in database, all modules available
            in database clustered under their functional <b>Title</b>
            categories.</li>
        </ol>
      </ol>
      <li>For help on <b>Lookup</b>, <b>Cells</b> and <b>Cytokines</b>,
        and for more detailed information: see <b>Instructions For Use</b>
        tab in <b>ReadMe</b>.<br>
      </li>
    </ol>
  </body>
</html>
